include/seqan3/search/detail/policy_max_error.hpp

```diff
@@ -24,44 +27,91 @@ namespace seqan3::detail
+                errors.deletion = get_error_count(get<search_cfg::max_error_deletion>(cfg).value, query_size);
+
+            errors.total = std::min<uint32_t>(255, errors.substitution + errors.insertion + errors.deletion);
+        }
+        else
+        { // set all error individually, if they have been set by the user
+            if constexpr (search_traits_t::has_max_error_total)
+                errors.total = get_error_count(get<search_cfg::max_error_total>(cfg).value, query_size);
+            if constexpr (search_traits_t::has_max_error_substitution)
+                errors.substitution = get_error_count(get<search_cfg::max_error_substitution>(cfg).value, query_size);
+            if constexpr (search_traits_t::has_max_error_insertion)
+                errors.insertion = get_error_count(get<search_cfg::max_error_insertion>(cfg).value, query_size);
+            if constexpr (search_traits_t::has_max_error_deletion)
+                errors.deletion = get_error_count(get<search_cfg::max_error_deletion>(cfg).value, query_size);
         }
 
-        return max_error;
+        return errors;
+    }
+
+private:
+
+    /*!\brief Returns a uint8_t object directly given by the search_cfg::error_count or calculated by the
+     *        search_cfg::error_rate depending on the given variant.
+     *
+     * \param[in] error_variant A search_cfg::error_count or search_cfg::error_rate.
+     * \param[in] query_size    The size of the query.
+     *
+     * \throws std::invalid_argument if error_rate is not between 0 and 1.
+     *
+     * \details If seqan3::search_cfg::error_count is set in the configuration, its value already is of type uint8_t and
+     *          can be returned as is.
+     *          If seqan3::search_cfg::error_rate is set in the configuration, the error rates are converted to error
+     *          counts based on the length of the query sequence.
+     */
+    uint8_t get_error_count(std::variant<search_cfg::error_count, search_cfg::error_rate> const & error_variant,
+                            [[maybe_unused]] size_t query_size)
+    {
+        return std::visit([&query_size] (auto error)
+            {
+                if constexpr (std::same_as<decltype(error), search_cfg::error_count>)
+                    return error.get();
+                else
+                {
+                    // check correct error rate values.
+                    if (0.0 > error.get()  || error.get() > 1.0)
+                        throw std::invalid_argument("Error rates must be between 0 and 1.");
+                    // make sure that error rate can be saved as uint8_t, so it is not too big in terms of query size
+                    const uint8_t& calculated_error_count = std::clamp(static_cast<int>(error.get() * query_size),
+                                                                       0,
+                                                                       INT8_MAX);
+                    return calculated_error_count;
```

Comment:
Both capture by reference is not save (it might be dangling, and a 64 bit reference is also bigger than an 8 bit integer), and the static cast to int is unsafe (might wrap around if result is bigger than what int can hold).

```suggestion
                    uint8_t const calculated_error_count = std::clamp(error.get() * query_size, 0.0, 255.0);
                    return calculated_error_count;
```